arch/x86/kvm/vmx/pkvm_init.c

```diff
@@ -102,12 +138,333 @@ static __init int pkvm_setup_host_vcpu(struct pkvm_hyp *pkvm, int cpu)
+	store_ldt(ldtr);
+	vmcs_write16(GUEST_LDTR_SELECTOR, ldtr);
+	vmcs_write32(GUEST_LDTR_AR_BYTES, 0x10000);
+	vmcs_writel(GUEST_LDTR_BASE, 0x0);
+	vmcs_write32(GUEST_LDTR_LIMIT, 0xffffffff);
+
+	/* Initialize IDTR */
+	store_idt(&dt);
+	vmcs_writel(GUEST_IDTR_BASE, dt.address);
+	vmcs_write32(GUEST_IDTR_LIMIT, dt.size);
+
+	/* Set MSRs */
+	vmcs_write64(GUEST_IA32_DEBUGCTL, 0);
+
+	rdmsrl(MSR_IA32_SYSENTER_CS, msrl);
+	vmcs_write32(GUEST_SYSENTER_CS, (u32)msrl);
+
+	rdmsrl(MSR_IA32_SYSENTER_ESP, msrl);
+	vmcs_writel(GUEST_SYSENTER_ESP, msrl);
+
+	rdmsrl(MSR_IA32_SYSENTER_EIP, msrl);
+	vmcs_writel(GUEST_SYSENTER_EIP, msrl);
+
+	rdmsrl(MSR_EFER, msrl);
+	vmcs_write64(GUEST_IA32_EFER, msrl);
+
+	rdmsrl(MSR_IA32_CR_PAT, msrl);
+	vmcs_write64(GUEST_IA32_PAT, msrl);
+}
+
+static __init void init_guest_state_area(void)
+{
+	init_guest_state_area_from_native();
+
+	/*Guest non register state*/
+	vmcs_write32(GUEST_ACTIVITY_STATE, GUEST_ACTIVITY_ACTIVE);
+	vmcs_write32(GUEST_INTERRUPTIBILITY_INFO, 0);
+	vmcs_writel(GUEST_PENDING_DBG_EXCEPTIONS, 0);
+	vmcs_write64(VMCS_LINK_POINTER, -1ull);
+}
+
+static __init void init_host_state_area(void)
+{
+	int cpu = smp_processor_id();
+	struct desc_ptr dt;
+	unsigned long msrl;
+	u16 selector;
+
+	vmcs_writel(HOST_CR0, read_cr0() & ~X86_CR0_TS);
+	/* Use host cr3 until the pkvm hypervisor created its own MMU */
+	vmcs_writel(HOST_CR3, __read_cr3());
+	vmcs_writel(HOST_CR4, __read_cr4());
+
+	/*
+	 * FIXME: Use the linux host environments for the pkvm hypervisor as
+	 * a temp solution before the isolation is fully functional.
+	 */
+	savesegment(cs, selector);
+	vmcs_write16(HOST_CS_SELECTOR, selector);
+	savesegment(ss, selector);
+	vmcs_write16(HOST_SS_SELECTOR, selector);
+	savesegment(ds, selector);
+	vmcs_write16(HOST_DS_SELECTOR, selector);
+	savesegment(es, selector);
+	vmcs_write16(HOST_ES_SELECTOR, selector);
+	savesegment(fs, selector);
+	vmcs_write16(HOST_FS_SELECTOR, selector);
+	vmcs_writel(HOST_FS_BASE, 0);
+	savesegment(gs, selector);
+	vmcs_write16(HOST_GS_SELECTOR, selector);
+	rdmsrl(MSR_GS_BASE, msrl);
+	vmcs_writel(HOST_GS_BASE, msrl);
+
+	vmcs_write16(HOST_TR_SELECTOR, GDT_ENTRY_TSS*8);
+	vmcs_writel(HOST_TR_BASE, (unsigned long)&get_cpu_entry_area(cpu)->tss.x86_tss);
+
+	native_store_gdt(&dt);
+	vmcs_writel(HOST_GDTR_BASE, dt.address);
+
+	store_idt(&dt);
+	vmcs_writel(HOST_IDTR_BASE, dt.address);
+
+	rdmsrl(MSR_IA32_SYSENTER_CS, msrl);
+	vmcs_write32(HOST_IA32_SYSENTER_CS, (u32)msrl);
+
+	rdmsrl(MSR_IA32_SYSENTER_ESP, msrl);
+	vmcs_writel(HOST_IA32_SYSENTER_ESP, msrl);
+
+	rdmsrl(MSR_IA32_SYSENTER_EIP, msrl);
+	vmcs_writel(HOST_IA32_SYSENTER_EIP, msrl);
+
+	rdmsrl(MSR_EFER, msrl);
+	vmcs_write64(HOST_IA32_EFER, msrl);
+
+	rdmsrl(MSR_IA32_CR_PAT, msrl);
+	vmcs_write64(HOST_IA32_PAT, msrl);
+}
+
+static __init void init_execution_control(struct vcpu_vmx *vmx)
+{
+	pin_controls_set(vmx, host_vmcs_config.pin_based_exec_ctrl);
+
+	/*
+	 * CR3 LOAD/STORE EXITING are not used by pkvm
```

Comment:
Late thought: worth adding a short note here that CR3 LOAD/STORE EXITING are a special case i.e. they are always read as 1 in MSR_IA32_VMX_PROCBASED_CTLS but are allowed to be cleared to 0 despite that?

BTW if I'm correctly reading SDM (Appendix A.3.2), looks like the fully correct way would be to check bit 55 in MSR_IA32_VMX_BASIC MSR, and then MSR_IA32_VMX_TRUE_PROCBASED_CTLS, to check if clearing CR3 LOAD/STORE EXITING is actually allowed? (Although maybe it's not worth extra complexity. If we ever encounter a platform where it is not allowed, IIUC we're gonna immediately notice that, since the host VM entry will fail. Also, the SDM also says that those IA32_VMX_TRUE_xxx are "new", and it's not clear how to check if the given CPU has those MSRs or not, before trying to read them...)

Reply:
Yes, you are right. KVM just clears these two bits directly, so probably we can do the same without introducing additional complexity. And we can add a short note for that.